File: Java/Plasma/PubUtilLib/plGImage/plFont.java
package Plasma.PubUtilLib.plGImage;

import java.io.IOException;
import java.io.FileNotFoundException;
import Plasma.CoreLib.hsStream;
import Plasma.NucleusLib.pnKeyedObject.*;

public class plFont extends hsKeyedObject {
    class FontLetterDef {
        public int offset;
        public int lHeight;
        public int stHeight;
        public float lPadding;
        public float rPadding;
    }

    protected String fontName;
    protected byte fontSize = -1;
    protected int fontFlags;
    protected int imgWidth;
    protected int imgHeight;
    protected int maxHeight;
    protected byte bpp;

    protected byte[] bmpData;
    protected short unknown;

    public FontLetterDef[] Letters = new FontLetterDef[256];
    
    public plFont() { }
    public plFont(String fileName) throws FileNotFoundException, IOException {
        hsStream FS = new hsStream(fileName, hsStream.MODE_READ);
        readP2F(FS);
    }

    public void read(hsStream S) throws IOException {
        super.read(S);
        readP2F(S);
    }

    public void write(hsStream S) throws IOException {
        super.write(S);
        writeP2F(S);
    }

    public void readP2F(hsStream S) throws IOException {
        fontName = S.readString(256);
        fontSize = S.readByte();
        fontFlags = S.readInt();
        imgWidth = S.readInt();
        imgHeight = S.readInt();
        maxHeight = S.readInt();
        bpp = S.readByte();

        bmpData = S.readString(imgWidth * imgHeight).getBytes();
        unknown = S.readShort();

        Letters = new FontLetterDef[S.readInt()];
        for (int i=0; i<Letters.length; i++) {
            Letters[i].offset = S.readInt();
            Letters[i].lHeight = S.readInt();
            Letters[i].stHeight = S.readInt();
            Letters[i].lPadding = S.readFloat();
            Letters[i].rPadding = S.readFloat();
        }
    }

    public void writeP2F(hsStream S) throws IOException {
        byte[] fName = new byte[256];
        System.arraycopy(fontName.getBytes(), 0, fName, 0, fontName.length());
        S.writeString(new String(fName));
        S.writeByte(fontSize);
        S.writeInt(fontFlags);
        S.writeInt(imgWidth);
        S.writeInt(imgHeight);
        S.writeInt(maxHeight);
        S.writeByte(bpp);

        S.writeString(new String(bmpData));
        S.writeShort(unknown);

        for (int i=0; i<Letters.length; i++) {
            S.writeInt(Letters[i].offset);
            S.writeInt(Letters[i].lHeight);
            S.writeInt(Letters[i].stHeight);
            S.writeFloat(Letters[i].lPadding);
            S.writeFloat(Letters[i].rPadding);
        }
    }
}

File: C++/PubUtilLib/plGImage/plFont.cpp
#include "plFont.h"
#include <string.h>

plFont::plFont(PlasmaVer pv) : FontSize(-1), FontFlags(0), imgWidth(0),
                   imgHeight(0), maxHeight(0), bpp(0), bmpData(NULL),
                   unknown(0), numLetters(256) {
    memset(FontName, '\0', 256);
    Letters = new FontLetterDef[numLetters];
}

plFont::plFont(const char* FileName, PlasmaVer pv)
       : hsKeyedObject::hsKeyedObject(pv) {
    hsStream *FS = new hsStream(FileName, fmRead);
    readP2F(FS);
    delete FS;
}

plFont::~plFont() {
    if (bmpData) delete bmpData;
    delete [] Letters;
}

FontLetterDef& plFont::operator[](int idx) {
    return Letters[idx];
}

void plFont::read(hsStream *S) {
    hsKeyedObject::read(S);
    readP2F(S);
}

void plFont::write(hsStream *S) {
    hsKeyedObject::write(S);
    writeP2F(S);
}

void plFont::readP2F(hsStream *S) {
    strncpy(FontName, S->readStr(256), 256);
    FontSize = S->readByte();
    FontFlags = S->readInt();
    imgWidth = S->readInt();
    imgHeight = S->readInt();
    maxHeight = S->readInt();
    bpp = S->readByte();

    bmpData = S->readStr((bpp * imgWidth * imgHeight) >> 3);
    unknown = S->readShort();

    numLetters = S->readInt();
    if (Letters) delete [] Letters;
    Letters = new FontLetterDef[numLetters];
    for (int i=0; i<numLetters; i++) {
        Letters[i].Offset = S->readInt();
        Letters[i].lHeight = S->readInt();
        Letters[i].stHeight = S->readInt();
        Letters[i].lPadding = S->readFloat();
        Letters[i].rPadding = S->readFloat();
    }
}

void plFont::writeP2F(hsStream *S) {
    S->writeStr(FontName, 256);
    S->writeByte(FontSize);
    S->writeInt(FontFlags);
    S->writeInt(imgWidth);
    S->writeInt(imgHeight);
    S->writeInt(maxHeight);
    S->writeByte(bpp);

    S->writeStr(bmpData, (bpp * imgWidth * imgHeight) >> 3);
    S->writeShort(unknown);

    S->writeInt(numLetters);
    for (int i=0; i<numLetters; i++) {
        S->writeInt(Letters[i].Offset);
        S->writeInt(Letters[i].lHeight);
        S->writeInt(Letters[i].stHeight);
        S->writeFloat(Letters[i].lPadding);
        S->writeFloat(Letters[i].rPadding);
    }
}

